lib/Dialect/Torch/IR/TorchOps.cpp

```diff
@@ -1393,6 +1393,48 @@ void AtenSortIntOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
   });
 }
 
+//===----------------------------------------------------------------------===//
+// AtenNewEmptyStridedOp
+//===----------------------------------------------------------------------===//
+
+void AtenNewEmptyStridedOp::getCanonicalizationPatterns(
+    RewritePatternSet &patterns, MLIRContext *context) {
+  patterns.add(+[](AtenNewEmptyStridedOp op, PatternRewriter &rewriter) {
+    SmallVector<int64_t> sizeListInts, strideListInts;
+    if (!matchPattern(op.getSize(), m_TorchListOfConstantInts(sizeListInts)))
+      return rewriter.notifyMatchFailure(
+          op, "all size list elements must be constant ints");
+    if (!matchPattern(op.getStride(),
+                      m_TorchListOfConstantInts(strideListInts)))
+      return rewriter.notifyMatchFailure(
+          op, "all stride list elements must be constant ints");
+
+    // We only support the cases with default stride values.
+    // For ex: aten.new_empty_strided(self, size=[2, 3, 4], stride=[12, 4, 1])
+    // Here the stride[0] == size[1] * size[2], stride[1] == size[2], and
+    // stride[2] == 1.
+    bool isDefaultStride = true;
+    for (unsigned i = 0; i < strideListInts.size(); i++) {
+      int64_t defaultStride = 1;
+      for (unsigned j = i + 1; j < sizeListInts.size(); j++)
+        defaultStride *= sizeListInts[j];
+      if (defaultStride != strideListInts[i]) {
+        isDefaultStride = false;
+        break;
+      }
+    }
+
+    if (!isDefaultStride)
+      return rewriter.notifyMatchFailure(
+          op, "only default strides supported for new_empty_strided op");
+
+    rewriter.replaceOpWithNewOp<AtenNewEmptyOp>(
```

Comment:
This is more a decomposition than a canonicalization. I would move it to decompose complex ops

Reply:
I was also confused between placing it as a decomposition or canonicalization. Thanks for the clarification! Btw, how do we primarily distinguish them?

Reply:
This is a pretty good article explaining canonicalizations: https://sunfishcode.github.io/blog/2018/10/22/Canonicalization.html. The main point being that a canonicalization makes "subsequent optimizations more effective." In this particular case, replacing `AtenNewEmptyStrided` with `AtenNewEmpty` does not really unlock new optimizations in torch-mlir, so it is more of a decomposition. Of course, the boundary between the canonicalizations and decompositions can often be very blurry. But in general in torch-mlir, if all the code does is replace one op with another, I would handle it as a decomposition.